mmcv/ops/cc_attention.py

```diff
@@ -2,91 +2,71 @@
-
-        ctx.save_for_backward(weight, v)
-
-        return out
-
-    @staticmethod
-    @once_differentiable
-    def backward(ctx, dout):
-        weight, v = ctx.saved_tensors
-        dw = torch.zeros_like(weight)
-        dv = torch.zeros_like(v)
-        ext_module.ca_map_backward(dout, weight, v, dw, dv)
-
-        return dw, dv
-
-
-ca_weight = CAWeightFunction.apply
-ca_map = CAMapFunction.apply
+    The diagonal are all "-inf". This is for avoiding calculating the
+    overlapped element in the Criss-Cross twice.
+    """
+    return torch.diag(torch.tensor(float('-inf')).to(device).repeat(n), 0)
 
 
 @PLUGIN_LAYERS.register_module()
 class CrissCrossAttention(nn.Module):
-    """Criss-Cross Attention Module."""
+    """Criss-Cross Attention Module.
+
+    Args:
+        in_channels (int): Channels of the input feature map.
+
+    Time consuming for one forward pass:
+        Settings:
+            Input size: [2,512,97,97]
+            Device: 1 NVIDIA GeForce RTX 2080 Ti
+        Results:
+            (with torch.no_grad())
+                The CUDA version: 0.0299619 s
+                This implementation: 0.00554402 s
+                Relative speed: 5.4x
+            (not with torch.no_grad())
+                The CUDA version: 0.0301349 s
+                This implementation: 0.00562803 s
+                Relative speed: 5.4x
+    """
 
     def __init__(self, in_channels):
-        super(CrissCrossAttention, self).__init__()
+        super().__init__()
         self.query_conv = nn.Conv2d(in_channels, in_channels // 8, 1)
         self.key_conv = nn.Conv2d(in_channels, in_channels // 8, 1)
         self.value_conv = nn.Conv2d(in_channels, in_channels, 1)
         self.gamma = Scale(0.)
         self.in_channels = in_channels
 
     def forward(self, x):
-        proj_query = self.query_conv(x)
-        proj_key = self.key_conv(x)
-        proj_value = self.value_conv(x)
+        """Forward pass of Criss-Cross Attention.
+
+        Args:
+            x (Tensor): Input feature map. \
+                shape (batch_size, in_channels, height, width)
+        Returns:
+            out (Tensor): shape (batch_size, in_channels, height, width)
+        """
+        B, C, H, W = x.size()
+        query = self.query_conv(x)
+        key = self.key_conv(x)
+        value = self.value_conv(x)
+        energy_H = torch.einsum('bchw,bciw->bwhi', query, key) + NEG_INF_DIAG(
+            H, query.device)
+        energy_H = energy_H.transpose(1, 2)
+        energy_W = torch.einsum('bchw,bchj->bhwj', query, key)
+        attn = F.softmax(
+            torch.cat([energy_H, energy_W], dim=-1), dim=-1)  # [B,H,W,(H+W)]
+        out = torch.einsum('bciw,bhwi->bchw', value, attn[..., :H])
+        out += torch.einsum('bchj,bhwj->bchw', value, attn[..., H:])
 
-        energy = ca_weight(proj_query, proj_key)
-        attention = F.softmax(energy, 1)
-        out = ca_map(attention, proj_value)
         out = self.gamma(out) + x
 
         return out
```

Comment:
The out tensor is not contiguous.Should it be contiguous to avoid error when out invokes view() method.

```python
cca = CrissCrossAttention(...)
input = ...
output = ccc(input)
output.view(-1)
```
The above code will raise an excption that `RuntimeError: view size is not compatible with input tensor's size and stride`

Reply:
The [unittest](https://github.com/open-mmlab/mmcv/blob/master/tests/test_ops/test_cc_attention.py) will fail if out is not contiguous.

Reply:
I didn't realize that. Fixed.

Reply:
https://github.com/open-mmlab/mmcv/blob/d3b057240829a0662125d5e83e002b33cd3fa874/tests/test_ops/test_cc_attention.py#L20-L21

The above lines can also be removed.

Reply:
ok